src/entities/user-with-email/user-with-email.ts

```diff
@@ -0,0 +1,43 @@
+import { Entity, UniqueEntityId } from '@/shared'
+import { Email, Password } from './value-objects'
+import { type EmailSignUpUserResponse, type EmailSignUpUserData } from './user-with-email-types'
+import { left, right } from '@/shared/either'
+
+export type UserWithEmailProps = {
+  email: Email
+  password: Password
+}
+
+export class UserWithEmail extends Entity<UserWithEmailProps> {
+  private constructor (props: UserWithEmailProps, id?: UniqueEntityId) {
+    super(props, id)
+    Object.freeze(this)
+  }
+
+  get email (): string {
+    return this.props.email.value
+  }
+
+  static register (data: EmailSignUpUserData): EmailSignUpUserResponse {
+    const { email, password } = data
+
+    const emailOrError = Email.create(email)
+    const passwordOrError = Password.create(password)
+
+    const results = [emailOrError, passwordOrError]
+
+    for (const result of results) {
+      if (result.isLeft()) return left(result.value)
+    }
+
+    return right(
+      new UserWithEmail(
+        {
+          email: emailOrError.value as Email,
```

Comment:
O uso do `as Email` para fazer uma type assertion pode ser problemático e deve ser evitado sempre que possível. O uso de `as` força o TypeScript a aceitar um tipo sem realizar a verificação, o que pode introduzir bugs difíceis de rasolver depois.

Uma maneira de evitar isso é garantir que o valor seja do tipo correto antes de ser usado. Como Email.create retorna um Either, o valor só deve ser acessado se o Either for right. Como o TypeScript já deve saber que o valor é do tipo Email quando você faz a verificação, o as Email não é necessário.

Algo como:

```
static register (data: EmailSignUpUserData): EmailSignUpUserResponse {
  const emailOrError = Email.create(data.email)
  if (emailOrError.isLeft()) return left(emailOrError.value)

  const passwordOrError = Password.create(data.password)
  if (passwordOrError.isLeft()) return left(passwordOrError.value)

  const email = emailOrError.value
  const password = passwordOrError.value

  return right(
    new UserWithEmail(
      { email, password },
      new UniqueEntityId(data.id)
    )
  )
}
```

Reply:
Entendido. Fiz essa refatoração.